Это вторая часть задания! 

Первая часть задания (задания 3.1-3.4) находится в [ноутбуке про линейную регрессию, который разбирали на лекции](https://colab.research.google.com/drive/1vLt8otyGyeiSIQ1XvQTXO9NozvBmqJ1q?usp=sharing).

# Предсказание цен на недвижимость

Компании по продаже недвижимости оценивают её стоимость, используя методы машинного обучения. 
Задача оценки цены на недвижимость также может быть нужна для выявления наиболее выгодных предложений, или, например, для выявления аномально низких цен на объект недвижимости.

**Постановка задачи анализа данных** 

Целью данной задачи является прогнозирование стоимости домов в округе Кинг (штат Вашингтон, США) с помощью построения регрессионных моделей и их анализа. Набор данных состоит из цен на дома в округе Кинг, проданных в период с мая 2014 года по май 2015 года. Данные опубликованы в открытом доступе на платформе Kaggle. 

Данные с сайта renthub.com по стоимости квартир для аренды:

<a href="https://drive.google.com/uc?id=1Dr-p-o5yMi3uIS2kPcUUwT5bH0cCt1Be
" target="_blank"><img src="https://drive.google.com/uc?id=1Dr-p-o5yMi3uIS2kPcUUwT5bH0cCt1Be" 
alt="IMAGE ALT TEXT HERE" width="360" border="0" /></a>


**Обзор доступных данных**

В наборе данных 21613 наблюдений и 16 признаков. 
Таким образом, про каждый из 21613 объектов недвижимости мы знаем значения 16 характеристик (число спален, оценка состояния риелтором, наличие вида на воду и т.п.)

Выборка была разбита на две части для обучения и для тестирования модели. Данные на обучение и тест обычно делят случайно и равномерно: нужно чтобы и обучающая, и тестовая выборка были похожие на те данные, на которых модель будет использоваться в боевых условиях. 

Более сложные разбиения могут использоваться если в данных есть временная компонента - например, мы строим скоринговую модель и для теста откладываем данные за последние несколько месяцев, а всю остальную выборку используем для обучения. Такое разбиение имитирует реальную работу модели: мы можем обучаться только на данных прошлых кредитов, а работать модель уже будет на данных из будущего. 

Доли общей выборки для обучения и тестирования обычно 70% и 30% соотвественно. Любые разумные числа подходят, если для обучения используется достаточно много данных (обычно больше 50%), но и для тестирования что-то остается (10% и больше).


Данные содержат два типа переменных:

* Целевая: **Целевая. Цена**
* Остальные переменные **могут использоваться для прогноза целевой переменной.**

## План анализа данных (data mining):

  1. Загрузить данные для обучения
  2. Обработать данные перед обучением модели
  3. Обучить модель на обучающей выборке
  4. Загрузить и предобработать данные для тестирования
  5. Провалидировать модель на тестовой выборке
  6. Проанализировать важность признаков и не забыть про постановку задачи

## 1. Загрузить данные для обучения

**Шаг 1.1. Загружаем библиотеки** 

Библиотека **warnings** отвечает за то, какие предупреждения (warnings) о работе будут выводиться пользователю. 
FutureWarning - предупреждения о том, как изменится работа библиотек в будущих версиях.
Поэтому такие предупреждения мы будем игнорировать.
Чтобы включить режим игнорирования мы отбираем все предупреждения из категории FutureWarning и выбираем для них действия 'ignore'.
Это делается вызовом функции simplefilter c задание двух атрибутов: действия action и категории предупреждений category.

In [ ]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

Для корректной работы с данными в python требуется загрузить специальную библиотеку
**pandas**, программную библиотеку на языке python для обработки и анализа данных. 

In [ ]:
import pandas as pd # загружаем библиотеку и для простоты обращения в коде называем её сокращенно pd

Для корректной работы с графиками в python требуется загрузить специальную библиотеку
**matplotlib**, программную библиотеку на языке python для визуализации данных двумерной и трехмерной графикой.

Графики используются для облегчения интерпретации полученных результатов, а также в качестве иллюстраций в презентациях и отчетах. 

Оснвные методы для построения:
* plot() - графики
* semilogy() - график логарифмический
* hist() - гистограммы

In [ ]:
import matplotlib.pyplot as plt # загружаем библиотеку и для простоты обращения в коде называем её сокращенно plt
# указываем, чтобы картинки отображались прямо в ноутбуке 
%matplotlib inline 

**Шаг 1.2. Загрузим данные**

Для решения задачи мы будем использовать данные. Они состоят из двух частей: часть для обучения и часть для тестирования модели. Загружаем данные с помощие команды !wget. Для того, чтобы игнорировать сообщения в процессе загрузки используем магическую команду %%capture в первой строке.

In [ ]:
%%capture
!wget https://www.dropbox.com/s/afwb0tnqm9izxha/predict_house_price_training_data.xlsx
!wget https://www.dropbox.com/s/sur2avqf4n5f4az/predict_house_price_test_data.xlsx

Так как данные в формате xlsx (Excel), мы будем использовать специальную функцию
из библиотеки pandas для загрузки таких данных **read_excel**.

В функции передаем один атрибут: название таблицы с данными.

In [ ]:
training_data = pd.read_excel('predict_house_price_training_data.xlsx') # загружаем таблицу в переменную training_data

*Что важно посмотреть после того, как мы загрузили данные?*
- проверить, что данные действительно загрузились
- посмотреть на данные, чтобы удостовериться, что они правильные: колонки имеют те же названия, что и в таблице и т.д.

Для того чтобы это сделать, нужно вызвать от переменной *training_data* метод **head()**, который выводит первые 5 строк таблицы. 

Для вызова метода объекта необходимо сначала написать *имя объекта*, затем поставить *точку*, затем уже написать *название метода*. Обратите внимание, что в конце обязательно ставить скобочки, потому что метод - это функция и в ней есть аргументы, просто в данном случае мы их не передаем, поэтому оставляем поле пустым

In [ ]:
training_data.head()

Целевая.Цена  Спальни  Ванные  ...  Год реновации   Широта  Долгота
0        830000        5    3.50  ...              0  47.6707 -122.144
1        385000        4    1.75  ...              0  47.5278 -122.345
2        610000        6    2.75  ...              0  47.6160 -122.115
3        550000        3    1.75  ...              0  47.5586 -122.173
4       1300000        3    2.75  ...              0  47.6389 -122.407

[5 rows x 16 columns]

**Шаг 1.3. Посмотрим на размеры загруженной таблицы**, у которой мы видели только первые 5 строк.

Для этого вызываем поле **shape** у нашей переменной *training_data*. Поле вызывается также как метод, но в конце скобки не ставятся, так как для поля не предусмотрена передача аргументов.  

In [ ]:
training_data.shape

(15129, 16)

*Что означает первое и второе число?*

Итак, таблица содержит 15129 строк (объектов) и 16 столбцов (признаков), включая выходной (целевой) признак.
15129 меньше 21613, так как мы пока загрузили только часть данных, которую будем использовать для обучения модели.

Таблицу проверили, теперь можно приступать к обработке данных.

# 2. Обработать данные перед обучением модели

**Шаг 2.1. Проверяем данные на наличие пропусков и типов переменных**

Начнем с проверки общей информации о данных.
Для того чтобы это сделать, нужно обратиться вызвать у переменной *training_data* метод **info()**.

Напомним, что в конце необходимо поставить скобочки.

In [ ]:
training_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15129 entries, 0 to 15128
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Целевая.Цена         15129 non-null  int64  
 1   Спальни              15129 non-null  int64  
 2   Ванные               15129 non-null  float64
 3   Жилая площадь        15129 non-null  int64  
 4   Общая площадь        15129 non-null  int64  
 5   Количество этажей    15129 non-null  float64
 6   Вид на воду          15129 non-null  int64  
 7   Просмотрены ранее    15129 non-null  int64  
 8   Состояние            15129 non-null  int64  
 9   Оценка риелтора      15129 non-null  int64  
 10  Площадь без подвала  15129 non-null  int64  
 11  Площадь подвала      15129 non-null  int64  
 12  Год постройки        15129 non-null  int64  
 13  Год реновации        15129 non-null  int64  
 14  Широта               15129 non-null  float64
 15  Долгота              15129 non-null 

Анализируем результата выполнения команды:

* 15129 строк (entries)
* 16 столбцов (Data columns)

В данных присутствует всего два типа dtypes:
* int64 - целое число  (12 столбцов)
* float64 - дробное число (4 столбца)

Цифры в каждой строчке обозначают количество заполненных (*non-null*) значений. Так как эти цифры в каждой строчке совпадают с числом строк (15129), то в данных нет пропусков и можно двигаться дальше. 

**Шаг 2.2. Работаем с целевой переменной**

*Какая переменная целевая?*

В данном случае по условию задачи мы должны прогнозировать стоимость, поэтому целевая переменная - это цена. 

In [ ]:
target_variable_name = 'Целевая.Цена'

Нам нужно выделить в отдельную переменную *training_values* столбец из нашей таблицы, который соответствует определенной выше целевой переменной. Для этого мы у таблицы *training_data* в квадратных скобках указываем имя нужного столбца. В нашем случае это имя записано в переменной *target_variable_name*. 

In [ ]:
training_values = training_data[target_variable_name]

Отделим входные переменные от выходной (целевой), чтобы можно было построить модель предсказания целевой переменной по входным. 
Для это нужно у переменной *training_data* вызвать метод **drop()**. Результат мы записываем в новую переменную *training_points*. После выполнения запроса *training_points* будет содержать исходную таблицу без целевого столбца. 

Обратите внимание, что в данном случае мы передаем два аргумента:
    1. target_variable_name - название столбца цены, который мы ранее записали в эту переменную и теперь хотим удалить из training_data
    2. axis=1 - означает, что мы удаляем столбец, а в случае axis=0 - означает, что мы удаляем строку

In [ ]:
training_points = training_data.drop(target_variable_name, axis=1)

Можно посмотреть результаты этих действий, вызвав метод **head()** и поле **shape**, которыми мы пользовались ранее, но сейчас нужно вызывать их от новой переменной *training_points*.

In [ ]:
training_points.head()

Спальни  Ванные  Жилая площадь  ...  Год реновации   Широта  Долгота
0        5    3.50           3490  ...              0  47.6707 -122.144
1        4    1.75           2360  ...              0  47.5278 -122.345
2        6    2.75           2040  ...              0  47.6160 -122.115
3        3    1.75           1940  ...              0  47.5586 -122.173
4        3    2.75           3450  ...              0  47.6389 -122.407

[5 rows x 15 columns]

In [ ]:
training_points.shape

(15129, 15)

Видно, что столбца действительно нет, а количество строк не изменилось. Данные в 5 первых строках такие же, как были ранее.

##   3. Обучить модель на обучающей выборке

**Шаг 3.1. Выбираем метод, который будем использовать**

Проще всего начать с простых методов. 

На выбор метода для построения модели влияет набор признаков, размер выборки, интуиция про то, какая связь между входными переменными и целевой. Но часто решение принимается исходя из того, какая модель сработала лучше.

Для корректной работы с методами построения моделей в python требуется загрузить специальную библиотеку
**sklearn**, программную библиотеку на языке python для для машинного обучения и анализа данных.

Импортируем модуль из этой библиотеки:
 * *linear_model* - тут находятся все линейные модели

In [ ]:
from sklearn import linear_model

Чтобы создать модель линейной регресии, пишем имя модуля 'linear_model', затем точку, затем название модели.

In [ ]:
linear_regression_model = linear_model.LinearRegression() # создаем модель

In [ ]:
linear_regression_model # смотрим, что получилось

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**Шаг 3.2. Обучить модель**

Теперь, когда мы создали прототип модели, можем её обучить на обучающей выборке. 

Для этого вызываем метод **fit()** и передаем ему на вход два аргумента: 
таблицу входных признаков и столбец значений целевой переменной - (training_points, training_values)

In [ ]:
linear_regression_model.fit(training_points, training_values)
linear_regression_model.coef_
# linear_regression_model.intercept_

array([-3.37568700e+04,  4.31210151e+04,  1.15347631e+02, -4.72733343e-02,
       -6.33671614e+03,  5.91890463e+05,  4.97598815e+04,  3.21627978e+04,
        9.96353171e+04,  7.99725606e+01,  3.53750698e+01, -2.44905828e+03,
        2.45690914e+01,  5.70765126e+05, -1.08281940e+05])

## Задание 3.5

Реализуйте настройку параметров модели не с помощью встроенной функции, а с помощью своей реализации.

Настроить параметры нужно с помощью **стохастического градиентного спуска**. 
В качестве дополнительного задания можно использовать регуляризацию **Elastic Net**. 

In [ ]:
# Calculate func, predict
def predict(row, coefficients, b):
  yhat = b
  for i in range(0, len(row)):
    yhat += coefficients[i] * row[i]
    # print(yhat)
  return yhat

def sgd(points, values, l_rate, n_epoch):
  coef = np.zeros(points.shape[1])
  b = 0
  for epoch in range(n_epoch):

    p = np.random.permutation(points.shape[0])
    points = points[p]
    values = values[p]

    for i in range(points.shape[0]):
      yhat = predict(points[i], coef, b)
      gw = 2*(yhat-values[i]) * points[i]
      gb = 2*(yhat-values[i])
      # print(yhat, values[i])
      coef = coef - (l_rate * gw)
      b = b - (l_rate * gb)

  return coef, b

## 4. Загрузить и предобработать данные для тестирования

**Шаг 4.1. Загрузим и проанализируем тестовые данные.**

Так как данные в формате xlsx (Excel), мы будем использовать специальную функцию
из библиотеки pandas для загрузки таких данных **read_excel**.

В функции передаем один атрибут: название файла, в котором находится таблица с данными.

In [ ]:
test_data = pd.read_excel('predict_house_price_test_data.xlsx')

*Что важно посмотреть, после того, как мы загрузили данные?*

- проверить, что данные действительно загрузились
- посмотреть на данные, чтобы удостовериться, что они правильные: колонки имеют те же названия, что и в таблице и т.д.

Для того чтобы это сделать, нужно вызвать от переменной *test_data* метод **head()**, который выводит первые 5 строк таблицы. 

Для вызова метода объекта необходимо сначала написать *имя объекта*, затем поставить *точку*, затем уже написать *название метода*. Обратите внимание, что в конце обязательно ставить скобочки, потому что метод - это функция и в ней есть аргументы, просто в данном случае мы их не передаем, поэтому оставляем поле пустым

In [ ]:
test_data.head()

Целевая.Цена  Спальни  Ванные  ...  Год реновации   Широта  Долгота
0        260000        3    1.00  ...           2007  47.3427 -122.087
1        734500        4    2.75  ...              0  47.7042 -122.107
2        325000        1    1.00  ...              0  47.4047 -122.331
3       1990000        3    2.50  ...              0  47.6281 -122.216
4        315000        3    2.00  ...              0  47.5374 -122.270

[5 rows x 16 columns]

Посмотрим на размеры загруженной таблицы, так как мы видели только 5 строк

Для этого вызываем поле **shape** у нашей переменной *test_data*. Поле вызывается также как метод, но в конце скобки не ставятся (!), так как для поля не предусмотрена передача аргументов.  

In [ ]:
test_data.shape

(6484, 16)

*Что означает первое и второе число?* 
Таблица содержит 6484 строк (объектов) и 16 столбцов (признаков), включая выходной (целевой) признак. Также как в учебных данных до обучения.

Таблицу проверили, теперь можно приступать к обработке данных
Действуем аналогично тому, как делали с данными для обучения

Проверим, есть ли в данных пропуски. Для того чтобы это сделать, нужно обратиться вызвать у переменной *test_data* метод **info()**.

In [ ]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6484 entries, 0 to 6483
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Целевая.Цена         6484 non-null   int64  
 1   Спальни              6484 non-null   int64  
 2   Ванные               6484 non-null   float64
 3   Жилая площадь        6484 non-null   int64  
 4   Общая площадь        6484 non-null   int64  
 5   Количество этажей    6484 non-null   float64
 6   Вид на воду          6484 non-null   int64  
 7   Просмотрены ранее    6484 non-null   int64  
 8   Состояние            6484 non-null   int64  
 9   Оценка риелтора      6484 non-null   int64  
 10  Площадь без подвала  6484 non-null   int64  
 11  Площадь подвала      6484 non-null   int64  
 12  Год постройки        6484 non-null   int64  
 13  Год реновации        6484 non-null   int64  
 14  Широта               6484 non-null   float64
 15  Долгота              6484 non-null   f

Цифры в каждой строчке обозначают количество заполненных (*non-null*) значений. Так как эти цифры в каждой строчке совпадают с числом строк (6484), то в данных нет пропусков.

**Шаг 4.2. Отделяем целевую переменную**

Нам нужно выделить в отдельную переменную *test_values* столбец из нашей таблицы, который соответствует определенной выше целевой переменной. Для этого мы у таблицы *test_data* в квадратных скобках указываем имя нужного столбца. В нашем случае это имя записано в переменной *target_variable_name*. 

In [ ]:
test_values = test_data[target_variable_name]

Отделим входные переменные от выходной (целевой), чтобы можно было построить модель предсказания целевой переменной по входным. 
Для это нужно у переменной *test_data* вызвать метод **drop()**. Результат мы записываем в новую переменную *test_points*. После выполнения запроса *test_points* будет содержать исходную таблицу без целевого столбца. 

Обратите внимание, что в данном случае мы передаем два аргумента:
    1. target_variable_name - название столбца цены, который мы ранее записали в эту переменную и теперь хотим удалить из training_data
    2. axis=1 - означает, что мы удаляем столбец, а в случае axis=0 - означает, что мы удаляем строку

In [ ]:
test_points = test_data.drop(target_variable_name, axis=1)

И проверяем результат записанный в test_points

In [ ]:
test_points.head()

Спальни  Ванные  Жилая площадь  ...  Год реновации   Широта  Долгота
0        3    1.00           1300  ...           2007  47.3427 -122.087
1        4    2.75           3280  ...              0  47.7042 -122.107
2        1    1.00           1220  ...              0  47.4047 -122.331
3        3    2.50           2880  ...              0  47.6281 -122.216
4        3    2.00           1300  ...              0  47.5374 -122.270

[5 rows x 15 columns]

In [ ]:
test_points.shape

(6484, 15)

Видно, что количество столбцов уменьшилось на один. Данные в 5 первых строках такие же, как были ранее.

# 5. Провалидировать модель на тестовой выборке

**Шаг 5.1. Сравнение моделей.**

Теперь мы готовы сравнить качество двух моделей! 😎

## Задание 3.6 Какая модель лучше?

Получим прогнозы целевой переменной на тестовых данных для модели линейной регрессии из sklearn и для своей реализации. 



In [ ]:
import numpy as np

In [ ]:
test_predictions_linear = linear_regression_model.predict(test_points)

In [ ]:
### введите свой код
### код для получения предсказаний с помощью своей реализации алгоритма настройки параметров модели
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error

# Min and max value for each row
def dataset_minmax(dataset):
	minmax = list()
	for i in range(len(dataset[0])):
		col_values = [row[i] for row in dataset]
		value_min = min(col_values)
		value_max = max(col_values)
		minmax.append([value_min, value_max])
	return minmax

# Each row in range from 0-1
def normalize_dataset(dataset, minmax):
	for row in dataset:
		for i in range(len(row)):
			row[i] = (row[i] - minmax[i][0]) / (minmax[i][1] - minmax[i][0])
   

dataset = np.array(training_points)
minmax = dataset_minmax(dataset)
normalize_dataset(dataset, minmax)

training_values = np.array(training_values)
training_points = np.array(training_points)

test_points_normalized = np.array(test_points)
minmax = dataset_minmax(test_points_normalized)
normalize_dataset(test_points_normalized, minmax)

In [ ]:
coef, bias = sgd(dataset, training_values, 0.00001, 100)

Iter:  0 Test MAE:  253824.28566511365 Train MAE  245251.06043180777
Iter:  1 Test MAE:  199117.96337900747 Train MAE  198224.64125684445
Iter:  2 Test MAE:  197134.72694867078 Train MAE  200194.15588637887
Iter:  3 Test MAE:  198835.50406921742 Train MAE  203891.17707471005
Iter:  4 Test MAE:  199403.4178757125 Train MAE  205548.50859899595
Iter:  5 Test MAE:  198581.87521618395 Train MAE  205262.9041032777
Iter:  6 Test MAE:  197484.85422334907 Train MAE  204602.41126317615
Iter:  7 Test MAE:  196142.54222219583 Train MAE  203599.70026277646
Iter:  8 Test MAE:  194718.27155897906 Train MAE  202467.14034077662
Iter:  9 Test MAE:  193180.25164889972 Train MAE  201156.02889420898
Iter:  10 Test MAE:  191871.8976357195 Train MAE  200151.5633245507
Iter:  11 Test MAE:  190208.36603357768 Train MAE  198637.68984763804
Iter:  12 Test MAE:  188730.17933777766 Train MAE  197363.74448145632
Iter:  13 Test MAE:  187399.80816099752 Train MAE  196287.85980223928
Iter:  14 Test MAE:  186020.073028

In [ ]:
sgd_model = linear_model.LinearRegression()
sgd_model.intercept_ = bias
sgd_model.coef_ = coef

Качество регрессионных моделей оценим двумя способами: 
1. Сравним визуально прогнозы с настоящими ценами (тестовые с предсказанием)
2. Сравним метрики качества MSE и MAE

Для корректного подсчета метрик качества модели в python требуется загрузить их из библиотеки **sklearn**. 

Мы используем две метрики качества:
 * *mean_absolute_error* - средняя абсолютная ошибка $|y_i - \hat{y}_i|$
 * *mean_squared_error* - средняя квадратичная ошибка $(y_i - \hat{y}_i)^2$

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

Подсчитаем ошибки для линейной модели.

Для этого вызовем методы **mean_absolute_error()** и **mean_squared_error()**. На вход им передается столбец настоящих значений *test_values* и столбец значений, предсказанных моделью линейной регрессии *test_predictions_linear*.

In [ ]:
mean_absolute_error_linear_model = mean_absolute_error(test_values, test_predictions_linear) 
mean_squared_error_linear_model = mean_squared_error(test_values, test_predictions_linear)

mean_absolute_error_sgd = mean_absolute_error(test_values, sgd_model.predict(test_points_normalized))
mean_squared_error_sgd = mean_squared_error(test_values, sgd_model.predict(test_points_normalized))

print("SGD: ")
print('Test MAE: ', mean_absolute_error_sgd)
print('Test RMSE: ', np.sqrt(mean_squared_error_sgd))
print("Lib: ")
print('Test MAE: ', mean_absolute_error_linear_model)
print('Test RMSE: ', np.sqrt(mean_squared_error_linear_model))

SGD: 
Test MAE:  132555.78484654933
Test RMSE:  247001.9860052878
Lib: 
Test MAE:  126852.51255005135
Test RMSE:  201883.24290317117


Подсчитайте ошибки для своей реализации модели.

Теперь напечатаем полученные ошибки. Обычно смотрят на корень из среднеквадратичной ошибки, RMSE. Чтобы извлечь корень нам понадобится библиотека **Numpy**. С помощью неё можно быстро производить вычисления сразу над массивами чисел.

## 6. Выявление важных признаков

Признаки, используемые для обучения модели, оказывают большое влияние на качество результатов. Неинформативные или слабо информативные признаки могут снизить качество модели, использование избыточных признаков в модели замедляет скорость ее работы.

Для того, чтобы отобрать признаки нам следует ответить на следующие вопросы: 

    1. Все ли признаки в наших данных заполненны разумными значениями?
    2. Какие признаки будут больше всего влиять на значение целевой переменной?
    3. Какие дополнительные признаки имело бы смысл добавить в список входных?

**6.1. Разглядывание значений признаков**

Для того чтобы ответить на первый вопрос, нам необходимо получить некоторую нашу экспертную оценку данных. В наших данных нет пропусков, однако иногда пропуски предварительно заполняют другими значениями (0, 999 и другие) и их можно увидеть, только внимательно посмотрев на данные. 

Для того, чтобы это сделать, нужно вызвать у переменной *training_points* метод **head(10)**, который выводит первые 10 строк таблицы.

In [ ]:
training_points = training_data.drop(target_variable_name, axis=1)
training_points.head(10)

Спальни  Ванные  Жилая площадь  ...  Год реновации   Широта  Долгота
0        5    3.50           3490  ...              0  47.6707 -122.144
1        4    1.75           2360  ...              0  47.5278 -122.345
2        6    2.75           2040  ...              0  47.6160 -122.115
3        3    1.75           1940  ...              0  47.5586 -122.173
4        3    2.75           3450  ...              0  47.6389 -122.407
5        3    2.25           2300  ...              0  47.5677 -122.086
6        3    2.50           2770  ...           2004  47.3685 -122.048
7        2    1.75           1650  ...              0  47.6871 -122.207
8        1    1.00            580  ...           2005  47.6829 -122.375
9        2    1.00            900  ...              0  47.5269 -122.314

[10 rows x 15 columns]

Рассмотрим столбец "Год реновации". Он содержит много нулей и мало реальных данных о годах, что как раз свидетельствует о том, что в этих данных изначально были пропуски, а к нам они попали уже заполненные нулями. 

Мы можем посмотреть количество уникальных значений и сколько раз эти значения встречаются в этом столбце. Для этого вызываем метод **value_counts()** у нашего столбца *training_points['Год реновации']*

In [ ]:
training_points['Год реновации'].value_counts()

0       14490
2014       63
2013       31
2000       28
2003       24
        ...  
1959        1
1948        1
1954        1
1962        1
1955        1
Name: Год реновации, Length: 67, dtype: int64

Как видно, нулей существенно больше (14490), чем любых других значений (639), поэтому можем сделать вывод, что этот признак не информативен и может негативно повлиять на качество модели.

**6.2. Какие признаки самые важные**

Ручной перебор признаков невозможен, если их тысячи, кроме того, часто даже эксперты не могут сказать, использование каких признаков в модели даст прирост качества.
Кроме ручного перебора всех признаков и их визуального анализа, можно использовать оценку значимости признаков с помощью обученных моделей. 

## Задание 3.7 

In [ ]:
## введите свой код
## напишите код, который вычисляет значимость признака (фактически это просто вес по модулю)
coef

array([  44011.00280319,  280201.16117028,  357782.52759812,
         16737.07398961,  168826.42890122,  159008.5414705 ,
        388487.41303873,   40660.42212106,  386438.70102609,
        379101.53122096,  198884.31576554, -111074.26169015,
        114354.95479184,  295936.07222209,  -21499.76253221])

Представьте полученные результаты в более удобном табличном виде. Создайте пустую таблицу *feature_importance* с двумя колонками "Название признака", "Важность признака". Для этого воспользуемся функцией **DataFrame()** из библиотеки pandas и передадим ей в качестве аргумента   
*columns = ["Название признака", "Важность признака"]*

In [ ]:
feature_importance = pd.DataFrame(columns = ['Название признака', 'Важность признака'])

Заполните соответствующие колонки значениями, которые мы получили выше. 

In [ ]:
# выведите полученную таблицу
feature_importance
training_points.columns[0]
# coef[0]
for i in range(training_points.columns.size):
  # df = pd.DataFrame([, ])
  feature_importance = feature_importance.append({'Важность признака': coef[i], 'Название признака': training_points.columns[i]}, ignore_index=True)

Мы можем отсортировать значения по важности, вызвав у *feature_importance* метод **sort_values()**.

Обратите внимание, что в данном случае мы передаем два аргумента:

1. by='Важность признака' - название столбца, по которому мы сортируем
2. ascending=False - означает, что мы сортируем по убиыванию, а в случае ascending=True сортировка будет по возростанию

In [ ]:
feature_importance.sort_values(by='Важность признака', ascending=False)

Название признака  Важность признака
6     Просмотрены ранее      388487.413039
8       Оценка риелтора      386438.701026
9   Площадь без подвала      379101.531221
2         Жилая площадь      357782.527598
13               Широта      295936.072222
1                Ванные      280201.161170
10      Площадь подвала      198884.315766
4     Количество этажей      168826.428901
5           Вид на воду      159008.541471
12        Год реновации      114354.954792
0               Спальни       44011.002803
7             Состояние       40660.422121
3         Общая площадь       16737.073990
14              Долгота      -21499.762532
11        Год постройки     -111074.261690

## Задание 3.8. 

Напомню исходную постановку задачи. Мы говорили о том, что оценка цены на недвижимость может быть нужна для выявления аномально низких цен на объект недвижимости. Давайте попробуем найти такие в датасете.

In [ ]:
predicted_values = sgd_model.predict(test_points_normalized)
actual_values = np.array(test_values)

In [ ]:
for i in range(actual_values.size):
  if actual_values[i] < predicted_values[i] * 0.5:
    print("Pred: ", predicted_values[i], "Actual: ", actual_values[i])

Pred:  418923.2344673465 Actual:  200000
Pred:  411734.5066272246 Actual:  175000
Pred:  1099542.6419344165 Actual:  425000
Pred:  880100.6491343939 Actual:  345600
Pred:  513551.51891351957 Actual:  225000
Pred:  602874.639400596 Actual:  285000
Pred:  369901.10025629814 Actual:  140000
Pred:  585639.9264115083 Actual:  270000
Pred:  633248.2680131489 Actual:  303500
Pred:  724673.1830807105 Actual:  350000
Pred:  886919.8581556265 Actual:  400000
Pred:  182424.35994103889 Actual:  83000
Pred:  478553.3129832053 Actual:  212000
Pred:  194835.74976327224 Actual:  92000
Pred:  388490.09749933495 Actual:  190500
Pred:  362886.92911773536 Actual:  170000
Pred:  352929.8238322818 Actual:  168500
Pred:  338612.2313904391 Actual:  152275
Pred:  884713.393417229 Actual:  357000
Pred:  609534.1754114706 Actual:  289275
Pred:  579288.9084310309 Actual:  285000
Pred:  1074849.5358023515 Actual:  400000
Pred:  532112.0912823449 Actual:  248000
Pred:  168791.098569009 Actual:  80000
Pred:  797676.